File: 2022/day-25/day-25.c
/* --- Advent of Code 2023 - Day 25: Full of Hot Air --- */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define MAX_LENGTH 256
#define BASE 5

/* solution part-1 */
int part_1(char input_file[]) {
    FILE *fp = fopen(input_file, "r");
    char buffer[MAX_LENGTH];
    long long base10_res = 0;

    /* file opened successfully */
    if (fp != NULL) {
        while (fgets(buffer, MAX_LENGTH - 1, fp) != NULL) {
            long long base10_sum = 0;

            /* convert SNAFU to base 10 */
            for (int pow_val = strlen(buffer) - 2; pow_val >= 0; pow_val--) {
                switch (buffer[strlen(buffer) - 2 - pow_val]) {
                    case '-':
                        base10_sum += pow(BASE, pow_val) * -1;
                        break;
                    case '=':
                        base10_sum += pow(BASE, pow_val) * -2;
                        break;
                    default:
                        base10_sum += pow(BASE, pow_val) * (buffer[strlen(buffer) - 2 - pow_val] - '0');
                }
            }

            base10_res += base10_sum;
        }

        /* convert base 10 result to base 5 */
        int base5_res[MAX_LENGTH];
        int len = 0;

        for (len = 0; base10_res > 0; len++) {
            base5_res[len] = base10_res % BASE;
            base10_res = base10_res / BASE;
        }

        /* adjust base 5 to SNAFU */
        int carry = 0;
        for (int i = 0; i < len; i++) {
            if ((base5_res[i] + carry) <= 2) {
                base5_res[i] = base5_res[i] + carry;
                carry = 0;
            } else {
                base5_res[i] = (base5_res[i] + carry) - 5;
                carry = 1;
            }
        }

        /* print result in SNAFU */
        for (int i = len - 1; i >= 0; i--) {
            switch (base5_res[i]) {
                case -1:
                    printf("-");
                    break;
                case -2:
                    printf("=");
                    break;
                default:
                    printf("%d", base5_res[i]);
            }
        }
        printf("\n");

    /* file not found */
    } else {
        fprintf(stderr, "Problems opening file '%s'\n", input_file);
        exit(1);
    }

    return 0;
}

/* solution part-2 */
int part_2(char input_file[]) {
    FILE *fp = fopen(input_file, "r");
    char buffer[MAX_LENGTH];

    /* file opened successfully */
    if (fp != NULL) {
        printf(":)\n");

        /* file not found */
    } else {
        fprintf(stderr, "Problems opening file '%s'\n", input_file);
        exit(1);
    }

    return 0;
}

/* main function */
int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s -p[1|2] <input-file>\n", argv[0]);
        return 1;
    }

    if (!strcmp(argv[1], "-p1")) {
        part_1(argv[2]);
    } else if (!strcmp(argv[1], "-p2")) {
        part_2(argv[2]);
    } else {
        fprintf(stderr, "Invalid argument: '%s'\n", argv[1]);
        return 1;
    }

    return 0;
}

File: 2022/day-25/day-25.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

""" --- Advent of Code 2023 - Day 25: Full of Hot Air --- """

import os
import sys

base = 5


# convert snafu to base10
def snafu_to_base10(snafu: str) -> int:
    base10_sum = 0

    for i, digit in enumerate(reversed(snafu)):
        match digit:
            case '-':
                base10_sum += base**i * -1
            case '=':
                base10_sum += base**i * -2
            case _:
                base10_sum += base**i * int(digit)

    return base10_sum


# convert base10 to base5
def base10_to_base5(base10: int) -> str:
    base5 = ""

    while base10 > 0:
        base5 += str(base10 % base)
        base10 //= base

    return base5


# adjust base5 to snafu
def base5_to_snafu(base5: str) -> str:
    snafu = ""
    carry = 0

    for digit in base5:
        if int(digit) + carry <= 2:
            snafu += str(int(digit) + carry)
            carry = 0
        else:
            match int(digit) + carry - base:
                case -1:
                    snafu += '-'
                case -2:
                    snafu += '='
                case _:
                    snafu += str(int(digit) + carry - base)

            carry = 1

    return snafu


# solution part 1
def part_1():
    base10_res = 0

    # read input file without newline and convert snafu to base10
    with open("input.txt", "r") as f:
        for line in [line.rstrip() for line in f.readlines()]:
            base10_res += snafu_to_base10(line)

    # convert base10 to base5 and then to snafu
    snafu_res = base5_to_snafu(base10_to_base5(base10_res))

    # print snafu backwards
    print(snafu_res[::-1])


# solution part 2
def part_2():
    print(":)")


# main entry point
os.chdir(os.path.dirname(os.path.abspath(__file__)))

if len(sys.argv) > 1 and sys.argv[1] == "-p1":
    part_1()
elif len(sys.argv) > 1 and sys.argv[1] == "-p2":
    part_2()
else:
    print("Usage: python", sys.argv[0], "[-p1|-p2]")
    sys.exit(1)
